README.md
# Carbonate

> "Pop bottles." *-- Birdman*

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/99e1654102b74d82a63505145334e7ed)](https://www.codacy.com/app/graphite-project/carbonate?utm_source=github.com&utm_medium=referral&utm_content=graphite-project/carbonate&utm_campaign=badger)
[![Build Status](https://travis-ci.org/graphite-project/carbonate.svg?branch=master)](https://travis-ci.org/graphite-project/carbonate)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bhttps%3A%2F%2Fgithub.com%2Fgraphite-project%2Fcarbonate.svg?type=shield)](https://app.fossa.io/projects/git%2Bhttps%3A%2F%2Fgithub.com%2Fgraphite-project%2Fcarbonate?ref=badge_shield)
[![codecov](https://codecov.io/gh/graphite-project/carbonate/branch/master/graph/badge.svg)](https://codecov.io/gh/graphite-project/carbonate)

Graphite clusters are pretty cool. Here are some primitive tools to help you manage your graphite clusters.

All of the tools support two common arguments; the path to a config file, and the name of the cluster. Using these tools alongside a config file that describes your graphite clusters you can build up scripts to manage your metrics. Some of the tools could easily be replaced with one-liners in shell, but exist here for convenience and readability. The goal is to provide fast, predictable utilities that can easily be composed into more advanced tooling.

## Install
[Carbonate is available from Python official third party repository](https://pypi.python.org/pypi/carbonate/0.2.1) (aka PyPi) and as such can be installed via regular Python package managers.
Note that you might have to install a python package manager (e.g. apt-get install python-setuptools on a ubuntu host)

```
pip install carbonate
```
## The Config

Carbonate expects a configuration file that defines the clusters in your environment. The default config file is located at `/opt/graphite/conf/carbonate.conf` or can be provided on the command line. The default cluster is named 'main'. Both defaults can be overridden by setting in the environment `CARBONATE_CONFIG` and `CARBONATE_CLUSTER` respectively.

```
[main]
DESTINATIONS = 192.168.9.13:2004:carbon01, 192.168.9.15:2004:carbon02, 192.168.6.20:2004:carbon03
REPLICATION_FACTOR = 2
SSH_USER = carbon

[agg]
DESTINATIONS = 192.168.9.13:2004:carbon01, 192.168.9.15:2004:carbon02, 192.168.6.20:2004:carbon03
RELAY_METHOD = aggregated-consistent-hashing
REPLICATION_FACTOR = 2
SSH_USER = carbon

[fnv]
DESTINATIONS = 192.168.9.13:2004:ba603c36342304ed77953f84ac4d357b, 192.168.9.15:2004:5dd63865534f84899c6e5594dba6749a, 192.168.6.20:2004:866a18b81f2dc4649517a1df13e26f28
REPLICATION_FACTOR = 2
SSH_USER = carbonate
HASHING_TYPE = fnv1a_ch
```

You should take care to match the list of destination IPs or hostnames to the nodes in your cluster (i.e. it should match with routing configuretion of your carbon relay). Order is important because of how the consistent hash ring is created.

You can configure the relay method to be one of "consistent-hashing" or "aggregated-consistent-hashing". If omitted, "consistent-hashing" is used by default. Use of "aggregated-consistent-hashing" usually requires a rules file to be provided to relevant commands.

The replication factor should match the replication factor for the cluster.

Also, you can choose to provide a SSH user that will be used when carbonate requires connecting to another node in the cluster to perform an operation. If this is not provided, then the current user executing the command will be chosen.

Finally, you can provide HASHING_TYPE of your cluster. Default is `carbon_ch`, also `fnv1a_ch` is supported. Please note that for using `fnv1a_ch` hashing you need `carbon` 1.0.2 or newer installed (or you need to use [carbon-c-relay](https://github.com/grobian/carbon-c-relay) relay instead).

## The Tools

### carbon-hosts

```
usage: carbon-hosts [-h] [-c CONFIG_FILE] [-C CLUSTER]

Return the addresses for all nodes in a cluster

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
```

### carbon-lookup

```
usage: carbon-lookup [-h] [-c CONFIG_FILE] [-C CLUSTER] [-s] METRIC

Lookup where a metric lives in a carbon cluster

positional arguments:
  METRIC                Full metric name to search for

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
  -a AGGREGATION_RULES, --aggregation-rules AGGREGATION_RULES
                        File containing rules used in conjunction with the
                        "aggregated-consistent-hashing" relay method (default:
                        /opt/graphite/conf/aggregation-rules.conf)
  -s, --short           Only display the address, without port and cluster
                        name (default: False)
```

### carbon-list

```
usage: carbon-list [-h] [-c CONFIG_FILE] [-C CLUSTER] [-d STORAGE_DIR]

List the metrics this carbon node contains

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
  -d STORAGE_DIR, --storage-dir STORAGE_DIR
                        Storage dir (default: /opt/graphite/storage/whisper)
```

### carbon-sieve

```
usage: carbon-sieve [-h] [-c CONFIG_FILE] [-C CLUSTER] [-f METRICS_FILE]
                    [-n NODE] [-I]

Given a list of metrics, output those that belong to a node

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
  -a AGGREGATION_RULES, --aggregation-rules AGGREGATION_RULES
                        File containing rules used in conjunction with the
                        "aggregated-consistent-hashing" relay method (default:
                        /opt/graphite/conf/aggregation-rules.conf)
  -f METRICS_FILE, --metrics-file METRICS_FILE
                        File containing metric names to filter, or '-' to read
                        from STDIN (default: -)
  -n NODE, --node NODE  Filter for metrics belonging to this node (default:
                        self)
  -I, --invert          Invert the sieve, match metrics that do NOT belong to
                        a node (default: False)
```

### carbon-sync

```
usage: carbon-sync [-h] [-c CONFIG_FILE] [-C CLUSTER] [-f METRICS_FILE] -s
                   SOURCE_NODE [-d STORAGE_DIR] [-b BATCH_SIZE]
                   [--source-storage-dir SOURCE_STORAGE_DIR]
                   [--rsync-options RSYNC_OPTIONS] [--rsync-disable-copy-dest]
                   [--tmpdir TMP_STAGING_DIR] [--rsync-max-retries MAX_RETRIES]
                   [--rsync-retries-interval SECONDS] [--dirty] [-l] [-o]

Sync local metrics using remote nodes in the cluster

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (env: CARBONATE_CONFIG) (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (env: CARBONATE_CLUSTER) (default: main)
  -f METRICS_FILE, --metrics-file METRICS_FILE
                        File containing metric names to filter, or '-' to read
                        from STDIN (default: -)
  -s SOURCE_NODE, --source-node SOURCE_NODE
                        Override the source for metrics data (default: None)
  -d STORAGE_DIR, --storage-dir STORAGE_DIR
                        Storage dir (default: /opt/graphite/storage/whisper)
  -b BATCH_SIZE, --batch-size BATCH_SIZE
                        Batch size for the rsync job (default: 1000)
  --source-storage-dir SOURCE_STORAGE_DIR
                        Source storage dir (default:
                        /opt/graphite/storage/whisper)
  --rsync-options RSYNC_OPTIONS
                        Pass option(s) to rsync. Make sure to use "--rsync-
                        options=" if option starts with '-' (default: -azpS)
  --rsync-disable-copy-dest
                        Avoid --copy-dest, transfer all whisper data between
                        nodes. (default: False)
  --rsync-max-retries RETRIES
                        Number of times rsync will attempt to copy each batch
                        of metrics before moving on. If all retry attempts are
                        unsuccessful, carbon-sync will write a file containing
                        the name of each metric in the failed batch so they can
                        be easily retried at a later time. (Default: 3)
  --rsync-retries-interval SECONDS
                        How long to wait in between each rsync retry attempt
                        (see --rsync-max-retries). (default: 5)
  -t TMP_STAGING_DIR, --tmpdir TMP_STAGING_DIR
                        Specify an alternate location in which the temporary
                        rsync staging dirs will be created. This can be useful
                        for large syncs where the default location (as chosen
                        by mkdtemp) resides on a filesystem that's too small
                        to store all the metrics being copied from the remote
                        host.
  --dirty               If set, don't clean temporary rsync directory
                        (default: False)
  -l, --lock            Lock whisper files during filling (default: False)
  -o, --overwrite       Write all non nullpoints from src to dst (default:
                        False)
```

### carbon-path

```
usage: carbon-path [-h] [-c CONFIG_FILE] [-C CLUSTER] [-f METRICS_FILE] [-r]
                   [-p] [-d STORAGE_DIR]

Transform metric paths to (or from) filesystem paths

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
  -f METRICS_FILE, --metrics-file METRICS_FILE
                        File containing metric names to transform to file
                        paths, or '-' to read from STDIN (default: -)
  -r, --reverse         Transform from file paths to metric paths (default:
                        False)
  -p, --prepend         Prepend storage dir to file paths (default: False)
  -d STORAGE_DIR, --storage-dir STORAGE_DIR
                        Whisper storage directory to prepend when -p given
                        (default: /opt/graphite/storage/whisper)
```

### carbon-stale

```
usage: carbon-stale [-h] [-c CONFIG_FILE] [-C CLUSTER] [-f METRICS_FILE] [-r]
                    [-d STORAGE_DIR] [-l HOURS] [-o HOURS] [-w] [-p]

Find and list potentially stale metrics.

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Config file to use (default:
                        /opt/graphite/conf/carbonate.conf)
  -C CLUSTER, --cluster CLUSTER
                        Cluster name (default: main)
  -f METRICS_FILE, --metrics-file METRICS_FILE
                        File containing metric names to transform to file
                        paths, or '-' to read from STDIN (default: -)
  -r, --reverse         Output metrics which are not stale instead (default:
                        False)
  -d STORAGE_DIR, --storage-dir STORAGE_DIR
                        Whisper storage directory to prepend when -p given
                        (default: /opt/graphite/storage/whisper)
  -l HOURS, --limit HOURS
                        Definition of staleness, in hours (default: 24)
  -o HOURS, --offset HOURS
                        Use a whisper data window ending HOURS ago (implies
                        -w) (default: 0)
  -w, --whisper         Use whisper data instead of filesystem stat() call
                        (default: False)
  -p, --paths           Print filesystem paths instead of metric names
                        (default: False)
```

### whisper-aggregate

```
usage: whisper-aggregate [-h] [-f METRICS_FILE] [-d STORAGE_DIR]

Set aggregation for whisper-backed metrics this carbon instance contains

optional arguments:
  -h, --help            show this help message and exit
  -f METRICS_FILE, --metrics-file METRICS_FILE
                        File containing metric names and aggregation modes, or
                        '-' to read from STDIN (default: -)
  -d STORAGE_DIR, --storage-dir STORAGE_DIR
                        Whisper storage directory (default:
                        /opt/graphite/storage/whisper)
```

### whisper-fill

```
usage: whisper-fill [-h] [-l] [-o] SRC DST

Backfill datapoints from one whisper file into another

positional arguments:
  SRC             Whisper source file
  DST             Whisper destination file

optional arguments:
  -h, --help      show this help message and exit
  -l, --lock      Lock whisper files during filling (default: False)
  -o, --overwrite  Write all non nullpoints from src to dst (default: False)
```

## Example usage

### Resync a node in a cluster

```
#!/bin/sh
#
# Resync a node from other nodes in the cluster
#

LOCAL_IP="$1"

for h in $(carbon-hosts) ; do
  (
    ssh $h -- carbon-list |
    carbon-sieve -n $LOCAL_IP |
    carbon-sync -s $h
  ) &
done
```

### Rebalance a cluster

```
#!/bin/sh
#
# Rebalance a cluster from one size to another. Remember to cleanup metrics
# that no longer belong when all nodes are rebalanced!
#

LOCAL_IP="$1"
OLD_CLUSTER="old"
NEW_CLUSTER="main"

for h in $(carbon-hosts -C "$OLD_CLUSTER") ; do
  ssh $h -- carbon-list |
  carbon-sieve -C "$NEW_CLUSTER" -n $LOCAL_IP |
  carbon-sync -s $h
done
```

### List metrics that don't belong

```
#!/bin/sh
#
# List metrics from disk that don't belong
#

LOCAL_IP="$1"

carbon-list | carbon-sieve -I -n $LOCAL_IP
```

### Listing metrics that have stopped updating

Metrics with whisper data that is entirely blank for the last 2 hours (perhaps
useful if you suspect issues with fs timestamps or carbon clients writing in 'the
future'):

```
carbon-list | carbon-stale --whisper --limit=2
```

Metrics whose metrics files appear untouched for 48 hours or more (functionally
identical to `find /your/data/dir -type f -mtime +2`):

```
carbon-list | carbon-stale --limit=48
```

More interesting is if you use ``carbon-stale``, then sieve to identify stale
metrics that don't belong here (vs un-stale metrics that *do* belong here but
are misreported in carbon-sieve due to things like doubled-up periods in metric
paths due to broken collectors. It's a thing.)

```
carbon-list | carbon-stale --limit=48 | carbon-sieve -I -n $LOCAL_IP
```

To print file paths for use with e.g. `xargs rm` or whatnot, use `-p`:

```
carbon-list | carbon-stale -p | xargs -n 100 rm
```


# License

The code is available under the MIT license.





CHANGELOG.md
## master (unreleased)
 - Nothing yet

## v1.1.9
 - Refactor log lines to use native interpolation (#129, @drawks)
 - Resolves issue #127 - python3 safe stdout handling (#128, @drawks)
 - Changed mutable default argument in test_fill.py and test_sync.py (#126, @aastha12)

## v1.1.8
 - support for Aggregated-Consistent Hash (#121 / #56, @deniszh / @klynch)
 - handle transient network failures and support custom staging dir path (#122, @ryangsteele)

## v1.1.7
 - fixes python3 TypeError (#113, @l4r-s)
 - Change write mode to non-binary. (#111, @hdost)
 -  Add python3 testing (#110, @hdost)
 - add codecov (#112, @piotr1212)

## v1.1.6
 - Python 3 support (PR#107, @piotr1212)
 - Use --copy-dest, enabling the rsync algorithm when copying from remote to staging (PR#106, @luke-heberling)
 - fix lint errors (PR#105, @YevhenLukomskyi)
 - specify long_description_content_type, so that package description is properly rendered on pypi.org (PR#104, @YevhenLukomskyi)

## v1.1.4
 - Use the scandir version of os.walk if possible (PR#99, @deejay1)
 - Include LICENSE in MANIFEST.in (PR#100, @deejay1)

## v1.1.3
 - Fixing carbon router hash (PR#93, @deniszh)
 - Adding LICENSE file to pypi package (PR#97, @deniszh)

## v1.1.2
 - carbon-sync: expose --overwrite to copy all non-null datapoints (PR#89, @jehiah)
 - fill: fix bugs causing some data points not to be copied (PR#90, @jehiah)

## v1.1.1
 - passing empty list for nodes seems to work (PR#87, @olevchyk)

## v1.0.2
 - (Experimental) support for fnv1a_ch hashing (PR#83, @deniszh)

## v1.0.0
 - Bumping version to 1.0.0 to match Graphite releases
 - unpin Twisted, remove py26 support (PR#78, @iksaif). Carbonate should still be Python 2.6 compatible, but newer Twisted is not
 - Fix reading whisper_lock_writes from cluster config in carbon_sync (PR#77, @justincmoy)

## v0.2.3

 - support new versions of carbon (PR#74 @wdauchy)
 - lock whisper files during carbon-sync and whisper-fill (PR#48 @fillipog, PR#71 @deniszh)
 - Fixes and optimizations for carbonate (PR#68 @iksaif, PR#70 @deniszh)
 - Make the heal process resilient (PR#42 @jjneely, PR#69 @deniszh)
 - Add option to follow symlinks (PR#65 @tmak0)
 - Do not overwrite target WSP file if source is corrupt (PR#52 @jjneely)
 - 'carbon-stale' tool to help find metrics with blank/missing data (PR#30 @bitprophet)
 - remove shebang, not a cli script (PR#55 @piotr1212)
 - mention environment variables in README.md (PR#53 @fillipog)
 - Fixup badge link in readme (PR#66 @dozoisch)
 - Add dirty mode for sync (PR#39 @iksaif)
 - Read config-file and cluster options from environ (PR#49 @fillipog)

## v0.2.2

- Load `/opt/graphite/lib` into python's path when executing any binstubs
- Bugfix in carbon-sieve where `--node` was not specified (@kamaradclimber)
- Bugfix in carbon-sieve where main cluster doesn't exist (@seanpquig)
- Add new carbon-path tool :sparkles: (@bitprophet)
- Bugfix in whisper-aggregate
- Make carbon-sieve filter correctly with long address formats (@tail)
- Add `-s` flag to carbon-lookup that displays IP/hostname only
- Support passing options to rsync during carbon-sync (@jphines)

## v0.2.1

- Improved CLI entry points
- Support for whisper 0.9.9
- Fix time.time() usage bug
- Error out if whisper-fill source/dest don't exist

## v0.2.0

- Initial release


carbonate/cli.py
import argparse
import errno
import logging
import os
import sys

from functools import partial
from time import time

from .aggregation import setAggregation, AGGREGATION
from .fill import fill_archives
from .list import listMetrics
from .lookup import lookup
from .sieve import filterMetrics
from .stale import data, stat
from .sync import run_batch
from .util import (
    local_addresses, common_parser, metric_to_fs, fs_to_metric,
    metrics_from_args,
)

from .config import Config
from .cluster import Cluster


STORAGE_DIR = '/opt/graphite/storage/whisper'


def carbon_hosts():
    parser = common_parser('Return the addresses for all nodes in a cluster')

    args = parser.parse_args()

    config = Config(args.config_file)
    cluster = Cluster(config, args.cluster)

    cluster_hosts = [d[0] for d in cluster.destinations]

    print("\n".join(cluster_hosts))


def carbon_list():
    parser = common_parser('List the metrics this carbon node contains')

    parser.add_argument(
        '-d', '--storage-dir',
        default=STORAGE_DIR,
        help='Storage dir')

    parser.add_argument(
        '-s', '--follow-sym-links',
        action='store_true',
        help='Follow sym links')

    args = parser.parse_args()

    try:
        for m in listMetrics(args.storage_dir, args.follow_sym_links):
            print(m)
    except IOError as e:
        if e.errno == errno.EPIPE:
            pass  # we got killed, lol
        else:
            raise SystemExit(e)
    except KeyboardInterrupt:
        sys.exit(1)


def carbon_lookup():
    parser = common_parser('Lookup where a metric lives in a carbon cluster')

    parser.add_argument(
        'metric', metavar='METRIC', nargs=1,
        type=str,
        help='Full metric name to search for')

    parser.add_argument(
        '-a', '--aggregation-rules',
        default='/opt/graphite/conf/aggregation-rules.conf',
        help='File containing rules used in conjunction with the ' +
             '"aggregated-consistent-hashing" relay method')

    parser.add_argument(
        '-s', '--short',
        action='store_true',
        help='Only display the address, without port and cluster name')

    args = parser.parse_args()

    config = Config(args.config_file)
    cluster = Cluster(config, args.cluster, args.aggregation_rules)

    results = lookup(str(args.metric[0]), cluster)

    if args.short:
        for i, _ in enumerate(results):
            results[i] = results[i].split(':')[0]

    print("\n".join(results))


def carbon_sieve():
    parser = common_parser(
        'Given a list of metrics, output those that belong to a node')

    parser.add_argument(
        '-f', '--metrics-file',
        default='-',
        help='File containing metric names to filter, or \'-\' ' +
             'to read from STDIN')

    parser.add_argument(
        '-a', '--aggregation-rules',
        default='/opt/graphite/conf/aggregation-rules.conf',
        help='File containing rules used in conjunction with the ' +
             '"aggregated-consistent-hashing" relay method')

    parser.add_argument(
        '-n', '--node',
        help='Filter for metrics belonging to this node. Uses the local ' +
             'addresses if not provided.')

    parser.add_argument(
        '-I', '--invert',
        action='store_true',
        help='Invert the sieve, match metrics that do NOT belong to a node')

    args = parser.parse_args()

    config = Config(args.config_file)
    cluster = Cluster(config, args.cluster, args.aggregation_rules)
    invert = args.invert

    metrics = metrics_from_args(args)

    if args.node:
        match_dests = [args.node]
    else:
        match_dests = local_addresses()

    try:
        for metric in metrics:
            m = metric.strip()
            for match in filterMetrics([m], match_dests, cluster, invert):
                print(metric.strip())
    except KeyboardInterrupt:
        sys.exit(1)


def carbon_sync():
    parser = common_parser(
        'Sync local metrics using remote nodes in the cluster'
        )

    parser.add_argument(
        '-f', '--metrics-file',
        default='-',
        help='File containing metric names to filter, or \'-\' ' +
             'to read from STDIN')

    parser.add_argument(
        '-s', '--source-node',
        required=True,
        help='Override the source for metrics data')

    parser.add_argument(
        '-d', '--storage-dir',
        default=STORAGE_DIR,
        help='Storage dir')

    parser.add_argument(
        '-b', '--batch-size',
        default=1000,
        help='Batch size for the rsync job')

    parser.add_argument(
        '--source-storage-dir',
        default=STORAGE_DIR,
        help='Source storage dir')

    parser.add_argument(
        '--rsync-options',
        default='-azpS',
        help='Pass option(s) to rsync. Make sure to use ' +
        '"--rsync-options=" if option starts with \'-\'')

    parser.add_argument(
        '--rsync-disable-copy-dest',
        default=False,
        action='store_true',
        help='Avoid --copy-dest, transfer all whisper data between nodes.')

    parser.add_argument(
        '--dirty',
        action='store_true',
        help="If set, don't clean temporary rsync directory")

    parser.add_argument(
        '-l', '--lock',
        default=False,
        action='store_true',
        help='Lock whisper files during filling')

    parser.add_argument(
        '-o', '--overwrite',
        default=False,
        dest='overwrite',
        action='store_true',
        help='Write all non nullpoints from src to dst')

    parser.add_argument(
        '-t', '--tmpdir',
        default=None,
        help='Specify where temporary rsync directories will be created')

    parser.add_argument(
        '--rsync-max-retries',
        type=int,
        default=3,
        help='Maximum number of rsync attempts for each batch of metrics')

    parser.add_argument(
        '--rsync-retries-interval',
        default=5,
        help='How long to wait (in seconds) between rsync retry attempts')

    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO)

    config = Config(args.config_file)

    user = config.ssh_user(args.cluster)
    remote_ip = args.source_node
    remote = "%s@%s:%s/" % (user, remote_ip, args.source_storage_dir)

    whisper_lock_writes = config.whisper_lock_writes(args.cluster) or \
        args.lock

    metrics_to_sync = []
    metrics = metrics_from_args(args)

    start = time()
    total_metrics = 0
    batch_size = int(args.batch_size)

    rsync_options = args.rsync_options
    if not args.rsync_disable_copy_dest:
        rsync_options += ' --copy-dest="%s"' % args.storage_dir

    for metric in metrics:
        total_metrics += 1
        metric = metric.strip()
        mpath = metric_to_fs(metric)

        metrics_to_sync.append(mpath)

        if total_metrics % batch_size == 0:
            print("* Running batch %s-%s"
                  % (total_metrics-batch_size+1, total_metrics))
            run_batch(metrics_to_sync, remote,
                      args.storage_dir, rsync_options,
                      remote_ip, args.dirty, lock_writes=whisper_lock_writes,
                      overwrite=args.overwrite, tmpdir=args.tmpdir,
                      rsync_max_retries=args.rsync_max_retries,
                      rsync_retries_interval=args.rsync_retries_interval)
            metrics_to_sync = []

    if len(metrics_to_sync) > 0:
        print("* Running batch %s-%s"
              % (total_metrics-len(metrics_to_sync)+1, total_metrics))
        run_batch(metrics_to_sync, remote,
                  args.storage_dir, rsync_options,
                  remote_ip, args.dirty, lock_writes=whisper_lock_writes,
                  overwrite=args.overwrite, tmpdir=args.tmpdir,
                  rsync_max_retries=args.rsync_max_retries,
                  rsync_retries_interval=args.rsync_retries_interval)

    elapsed = (time() - start)

    print("")
    print("* Sync Report")
    print("  ========================================")
    print("  Total metrics synced: %s" % total_metrics)
    print("  Total time: %ss" % elapsed)


def carbon_path():
    # Use common_parser for consistency, even though we do not use any config
    # file options at present.
    parser = common_parser(
        'Transform metric paths to (or from) filesystem paths'
    )

    parser.add_argument(
        '-f', '--metrics-file',
        default='-',
        help='File containing metric names to transform to file paths, or ' +
        '\'-\' to read from STDIN')

    parser.add_argument(
        '-r', '--reverse',
        action='store_true',
        help='Transform from file paths to metric paths'
    )

    parser.add_argument(
        '-p', '--prepend',
        action='store_true',
        help='Prepend storage dir to file paths')

    parser.add_argument(
        '-d', '--storage-dir',
        default=STORAGE_DIR,
        help='Whisper storage directory to prepend or strip')

    args = parser.parse_args()
    metrics = metrics_from_args(args)
    if args.reverse:
        # Always try to strip storage dir when doing fs->metric.
        # Nobody would ever not want that.
        func = partial(fs_to_metric, prepend=args.storage_dir)
    else:
        # Only give non-empty prepend if -p given, when metric->fs
        prepend = args.storage_dir if args.prepend else None
        func = partial(metric_to_fs, prepend=prepend)

    for metric in metrics:
        print(func(metric))


def carbon_stale():
    # Use common_parser for consistency, even though we do not use any config
    # file options at present.
    parser = common_parser(
        'Find and list potentially stale metrics.'
    )

    parser.add_argument(
        '-f', '--metrics-file',
        default='-',
        help='File containing metric names to scan for staleness, or ' +
        '\'-\' to read from STDIN')

    parser.add_argument(
        '-r', '--reverse',
        action='store_true',
        help='Output metrics which are not stale instead')

    parser.add_argument(
        '-d', '--storage-dir',
        default=STORAGE_DIR,
        help='Whisper storage directory to prepend when -p given')

    parser.add_argument(
        '-l', '--limit', metavar='HOURS',
        type=int, default=24,
        help='Definition of staleness, in hours')

    parser.add_argument(
        '-o', '--offset', metavar='HOURS',
        type=int, default=0,
        help='Use a whisper data window ending HOURS ago (implies -w)')

    parser.add_argument(
        '-w', '--whisper',
        action='store_true',
        help='Use whisper data instead of filesystem stat() call')

    parser.add_argument(
        '-p', '--paths',
        action='store_true',
        help='Print filesystem paths instead of metric names')

    args = parser.parse_args()
    metrics = metrics_from_args(args)
    prefix = args.storage_dir
    use_whisper = args.whisper or args.offset
    for path in map(partial(metric_to_fs, prepend=prefix), metrics):
        passed = (data if use_whisper else stat)(path, args.limit, args.offset)
        value = path if args.paths else fs_to_metric(path, prepend=prefix)
        if (not passed) if args.reverse else passed:
            print(value)


def whisper_aggregate():
    parser = argparse.ArgumentParser(
        description='Set aggregation for whisper-backed metrics this carbon ' +
                    'instance contains',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument(
        '-f', '--metrics-file',
        default='-',
        help='File containing metric names and aggregation modes, or \'-\' ' +
             'to read from STDIN')

    parser.add_argument(
        '-d', '--storage-dir',
        default=STORAGE_DIR,
        help='Whisper storage directory')

    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO)

    metrics = metrics_from_args(args)

    metrics_count = 0

    for metric in metrics:
        try:
            name, t = metric.strip().split('|')

            mode = AGGREGATION[t]
            if mode is not None:
                path = metric_to_fs(name, prepend=args.storage_dir)
                metrics_count = metrics_count + setAggregation(path, mode)
        except ValueError as exc:
            logging.warning("Unable to parse '%s' (%s)" % (metric, str(exc)))

    logging.info('Successfully set aggregation mode for ' +
                 '%d of %d metrics' % (metrics_count, len(metrics)))


def whisper_fill():
    parser = argparse.ArgumentParser(
        description='Backfill datapoints from one whisper file into another',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument(
        'source',
        metavar='SRC',
        help='Whisper source file')

    parser.add_argument(
        'dest',
        metavar='DST',
        help='Whisper destination file')

    parser.add_argument(
        '-l', '--lock',
        default=False,
        action='store_true',
        help='Lock whisper files during filling')

    parser.add_argument(
        '-o', '--overwrite',
        default=False,
        dest='overwrite',
        action='store_true',
        help='Write all non nullpoints from src to dst')

    args = parser.parse_args()

    src = args.source
    dst = args.dest

    if not os.path.isfile(src):
        raise SystemExit('Source file not found.')

    if not os.path.isfile(dst):
        raise SystemExit('Destination file not found.')

    startFrom = time()

    fill_archives(src, dst, startFrom, lock_writes=args.lock,
                  overwrite=args.overwrite)
